### ISSUE TITLE
Task.Delay/Task.Yield Results in Unresponsive Blazor Application on Callback (`@onclick`)

### ISSUE THREAD

**Issue by Mike-E-angelo** [dotnet/aspnetcore#36799]
### Describe the bug
Blazor will successfully execute an asynchronous method containing `Task.Delay` or `Task.Yield` on pre-render and render, but silently fails when executed on a event callback such as one invoked via an `@onclick` attribute.

When failing, the application becomes unresponsive without any exception thrown, and a page refresh is required.

### To Reproduce

I have been spending the past several days trying to track this down, and have boiled it into a simple SLN which you can find here:
https://github.com/Mike-E-angelo/Stash/tree/master/Blazor.DelayYield.Unresponsive

Clicking on the `Click Me To Break` will call the `Refresh` method, and will [go as far as this line](https://github.com/Mike-E-angelo/Stash/blob/master/Blazor.DelayYield.Unresponsive/Blazor.DelayYield.Unresponsive/Pages/Index.razor#L47), but after this, the application becomes unresponsive without a thrown exception for explanation.

### Exceptions (if any)

None, unfortunately.  One would be nice. :)

### Further technical details
- ASP.NET Core version `6.0.0-rtm.21471.8`
- Visual Studio 2022 Preview 4.1
```
> dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.100-rtm.21466.6
 Commit:    448400e0a5

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-rtm.21466.6\

Host (useful for support):
  Version: 6.0.0-rc.2.21461.1
  Commit:  71fe9fb8ec

.NET SDKs installed:
  3.1.302 [C:\Program Files\dotnet\sdk]
  3.1.412 [C:\Program Files\dotnet\sdk]
  6.0.100-preview.7.21379.14 [C:\Program Files\dotnet\sdk]
  6.0.100-rc.1.21463.6 [C:\Program Files\dotnet\sdk]
  6.0.100-rtm.21466.6 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview7.19365.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.1.21452.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.2.21462.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview5-27626-15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.1.21451.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.2.21461.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.1.21451.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.2.21461.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by javiercn** [comment]
@Mike-E-angelo thanks for contacting us.

I've looked a bit at what you are doing on the repo and there seems to be a lot of code involving your own task management.

If you remove all your code and do a `Task.Yield`, does it break?

**Comment by Mike-E-angelo** [comment]
Thank you for your quick reply and valuable feedback @javiercn.

I took your feedback and investigated further.

If I reduce everything to a `Func<Task<string>>`, then it works as expected on a callback.  However, if I [utilize a simple wrapper around this delegate](https://github.com/Mike-E-angelo/Stash/blob/60b89f862a1b13d55be0ed183c6208489c96af85/Blazor.DelayYield.Unresponsive/RazorClassLibrary1/Class1.cs#L7-L14), then that is when it breaks.

(I've updated the code to remove everything but the control and simple wrapper mentioned above.  I am open to more feedback as required to further diagnose this issue.)

Two concerns here:

1. As mentioned, this works as expected for initial rendering and pre-rendering.  It isn't until a delegate/callback/`@onclick` is invoked that this problem is presented.
2. There is no exception being thrown, not even a first-chance error message in the Debug window.  The user-scoped session simply stops working without any sort of clue/feedback why.

Finally, I also took some time to check that this also occurs in `net5.0` as well.

**Comment by javiercn** [comment]
@Mike-E-angelo there's a lot of code in your sample and it's not clear what you are trying to accomplish nor that its a framework bug.

For us to make progress in this issue, we need a truly minimal repro that doesn't involve all the machinery you are adding on your sample, otherwise we don't know if we are debugging an issue in the framework or in your application's code.

I would suggest you add code bit by bit until you are able to trigger the issue and explain how the framework is not behaving correctly. Otherwise, we can't tell if its the framework or your app.

**Comment by Mike-E-angelo** [comment]
I am super appreciative of your patience here, @javiercn.  It's been a long 48 hours for me.  I found the problem and it was hidden away in a pattern match.  At one point I had that set to `IsCompleted` but sometime during my extensive editing (and re-editing) [I moved that to the `Result` property ](https://github.com/Mike-E-angelo/Stash/blob/master/Blazor.DelayYield.Unresponsive/RazorClassLibrary1/OperationContentContainer.razor#L16) which as you know is a massive no-no in `await/async`.

What was getting to me is that it was working pre-rendering/rendering, and only failing in the `@onclick` delegate.  I now know that a deadlock was occurring, which of course means no exception can be thrown.

Ouch.  Painful lesson.  I further now know that anytime the application inexplicably stops working as it did here that it's an indication of a deadlock somewhere.  Well, I _did_ know that (and previously looked) that but I sorta hung myself this time with the pattern matching.

Anyways, thanks again for taking the time to walk me through that.  Much appreciation for all the great work you are doing out there, including constantly dealing with runts like me that think there's a problem somewhere. 😅
